worldeditadditions/utils/parse/axes_parser.lua

```diff
@@ -0,0 +1,204 @@
+end
+
+local key_instance
+if worldeditadditions then
+	local wea = worldeditadditions
+	key_instance = dofile(wea.modpath.."/utils/parse/key_instance.lua")
+else
+	key_instance = require("worldeditadditions.key_instance")
+end
+
+--- Unified Axis Keywords banks
+local keywords = {}
+-- Direction keywords
+keywords.dir = {
+	["?"] = "front", f = "front",
+	facing = "front", front = "front",
+	b = "back", back = "back",
+	behind = "back", rear = "back",
+	l = "left", left = "left",
+	r = "right", right = "right",
+	u = "up", up = "up",
+	d = "down", down = "down",
+}
+-- Mirroring keywords
+keywords.mirroring = {
+	sym = true, symmetrical = true,
+	mirror = true, mir = true,
+	rev = true, reverse = true,
+	["true"] = true
+}
+
+--- Initialize parser function container
+local parse = {}
+
+--- Processes an axis declaration into ordered xyz format. (Supports axis clumping)
+-- For example, "zzy" would become "yz"
+-- @param: str: String: Axis declaration to parse
+-- @returns: Table|Bool: axis | axes | false
+function parse.axis(str)
+	local axes, ret = {"x","y","z"}, {}
+	for i,v in ipairs(axes) do
+		if str:match(v) then table.insert(ret,v) end
+	end
+	if #ret > 0 and str:match("^[xyz]+$") then
+		return ret
+	elseif str == "h" then
+		return {"x", "z"}
+	elseif  str == "v" then
+		return {"y"}
+	else return false end
+end
+
+--- Processes an number from a string.
+-- @param: str: String: string to parse
+-- @returns: Number|Bool: processed number | false
+function parse.num(str)
+	str = tostring(str) -- To prevent meltdown if str isn't a string
+	local num = str:match("^-?%d+%.?%d*$")
+	if num then
+		return tonumber(num)
+	else return false end
+end
+
+--- Checks if a string is a valid Unified Axis Keyword. (Supports axis clumping)
+-- @param: str: String: Keyword instance to parse
+-- @returns: Key Instance: returns keyword type, processed keyword content and signed number (or nil)
+function parse.keyword(str)
+	if type(str) ~= "string" then
+		return key_instance.new("err", "Error: \""..tostring(str).."\" is not a string.", 404)
+	end
+	local sign = 1
+	if str:sub(1,1) == "-" then
+		sign = -1
+		str = str:sub(2)
+	end
+	
+	local axes = parse.axis(str)
+	if axes then
+		return key_instance.new("axis", axes, sign)
+	elseif keywords.dir[str] then
+		return key_instance.new("dir", keywords.dir[str], sign)
+	elseif keywords.mirroring[str] then
+		return key_instance.new("rev", "mirroring")
+	else return key_instance.new("err", "Error: \""..str.."\" is not a valid axis, direction or keyword.", 422)
+	end
+end
+
+--- Creates a vector with a length of (@param: value * @param: sign)
+--  on each axis in @param: axes.
+-- @param: axes: Table: List of axes to set
+-- @param: value: Number: length of to set axes
+-- @param: sign: Number: sign multiplier for axes
+-- @returns: Vector3: processed vector
+function parse.vectorize(axes,value,sign)
+	local ret = Vector3.new()
+	for i,v in ipairs(axes) do
+		ret[v] = value * sign
```

Comment:
This is confusing. `ref` is a `Vector3` here, which looks like `{ x = 4, y = 5, z = 6 }`. Yet this appears to use the Vector3 instance as a list? I'm confused.

Reply:
1. It's `ret` (short for return) not `ref`
2. It's not using the `Vector3` as a list but rather populating a blank one axis by axis with the input value. (This was the fastest way to do it.)

Reply:
Ahh I see! Thanks for claarifying.

Oops, yes that's a typo in my comment there.